murdock: implement numeric priorities

kaspar030 opened the pull request:
This PR drops the fasttrack queue and turns the "regular" queue into a priority queue.
It adds configurable priority modifiers (per-branch, per-label) that get applied to queued or (re-)labeled jobs.

Previously, it was possible to "fasttrack" jobs, moving them ahead of "regular" jobs, essentially supporting two priorities.
With this PR, the job priority becomes numeric. This is limiting.

As an actual use-case, RIOT will move to bors for builds that should get merged, but non-bors quickbuilds for PRs.
The bors builds are expected to take >2hrs, whereas the quickbuilds are below 5min.
PR builds should be prioritized over the long bors builds.
Up to this point, two priorities work fine (fasttrack all PR builds).

Now it would be great if _some_ PR builds could be set to a lower priority (for contributors that want to use CI for build testing without hampering workflow), or even higher priority (for important backports or CI fixes).
Also, it would be great if e.g., adding the "CI: high priority") label would update the job's queue position.
And, if "can_fast_ci_build.py" determines a subset build, the priority can be dynamically adapted.

So this PR implements the following changes:

- [x] replace the fasttrack and regular queues with one priority queue
- [x] adds a "priority" field to MurdockJob
- [x] add an `__lt()` implementation to MurdockJob (making it sort by priority)
- [x] make MurdockJob calculate it's priority based on `priorities: ` repo-level (`.murdock.yml`) configuration
- [x] adds `priorities:` `.murdock.yml` configuration
- [x] implements requeuing on label change (if the priority changes)

bergzand commented on murdock/murdock.py:
```diff
@@ -210,13 +214,17 @@ async def add_job_to_queue(self, job: MurdockJob):
-            self.queue.put_nowait(job)
+            prio = 0
+
+        # PriorityQueue is a min_queue, but "higher priority equals higher value"
+        # makes more sense. So, invert the priority by multiplying with `-1`
+        self.queue.put_nowait(PrioritizedItem(prio * -1, job))
```
I suggest wrapping this into a separate function. The `prio * -1` makes this error prone when copy-pasted around if used elsewhere.

```Python
def _submit(self, job: Any, prio: int):
    """
    Submit a job into the priority queue. Jobs with a high prio value are served before jobs with low prio value.

    :param job: Job to submit
    :param prio: Priority of the job
    """
    self.queue.put_nowait(PrioritizedItem(prio * -1, job))
```

kaspar030 replied:
would that be a function level function? or class level?

bergzand replied:
@aabadie Any opinion here?

aabadie replied:
If the priority is added as an attribute to MurdockJob, there won't be a need for wrapper item. The `__le__` method of MurdockJob could take care of inverting the order.

kaspar030 commented on murdock/murdock.py:
```diff
@@ -46,6 +47,14 @@
+
+
+@dataclass(order=True)
+class PrioritizedItem:
```
ok then, I'll add the `__le__()`.

kaspar030 commented:
Ok, I've added a priority field to MurdockJob.
I've left the logic on how to set the priority value a bit open, there's now a `calculate_priority()` static_method on MurdockJob, which currently only takes the fasttrack label into account.

We can discuss whether fasttracking is needed at all, but maybe it's nice to have it as "build that next" hammer.

@aabadie how does the added JobModel field interact with the existing database?

kaspar030 commented:
adding the priority configuration was straight-forward.
I might need some help with adding tests, I don't understand the test suite (yet).

aabadie commented on murdock/job.py:
```diff
@@ -51,6 +52,9 @@ def __init__(
             else False
         )  # type: ignore[union-attr]
+        self.priority: int = MurdockJob.calculate_priority(
+            config, pr, ref, self.fasttracked
```
all these function parameters are already instance attributes. Thus there's no need for `calculate_priority` to be a staticmethod. A regular instance method, private, should be enough `def _priority(self):` and use `self.config`, `self.pr`, `self.ref` and `self.fastracked` inside.

aabadie replied:
Note that an `@property` method could also be used but it would recompute the priority each time it's accessed. Maybe that better I don't know

kaspar030 replied:
I wasn't sure, can safely use a class method within `__init__()`?
(Also it needs to be public so it can react to label changes.)

aabadie replied:
> can safely use a class method within __init__()?

yes, if the method is accessing already defined/initialized attributes.

>  it needs to be public so it can react to label changes

then I think a property method is more appropriate.

kaspar030 replied:
a property method would make e.g., "job.priority" actually call the method?

aabadie replied:
yes so each time `job.priority` is accessed the new value is automatically recomputed

kaspar030 replied:
In that case I'd rather have a class method.
like, if a label changes, I'd get the job object, get it's current priority, get calculate the priority it would have taking the new labels into account, and if they don't match, re-queue:

```
if job.calculate_priority() != job.priority:
  ... requeue
```

kaspar030 replied:
assuming the pr field of a job is a reference to updated PR data.
if that's not the case, it's easier to have this staticmethod and supply all needed data at the call site.

aabadie replied:
How do you requeue ? In the current state, a new MurdockJob instance is created and the previous matching ones are marked as cancelled.

I also thought that the display order must also be taken into account. This is something that is in another data structure, a simple list. So a sort on priority must be added there.

https://github.com/murdock-ng/murdock/blob/e62397d29e8ffe1ded2bbb2dd0a7f7a0c49ee853/murdock/murdock.py#L561-569

kaspar030 replied:
I would add re-queueing here:
https://github.com/murdock-ng/murdock/blob/e62397d29e8ffe1ded2bbb2dd0a7f7a0c49ee853/murdock/murdock.py#L478

(if the job is queued, re-calc its priority and compare to the queued one. if it has changed, don't "return" but fall through to "schedule_job()", which would cancel the old job and create a new, sorted into the queue by priority.)

aabadie replied:
I'm a bit lost. What is the use case exactly ?

aabadie commented:
> I might need some help with adding tests, I don't understand the test suite (yet).

The test suite is becoming huge I admit. Maybe start by adding a test function in `test_job.py` to check the `job.priority` attribute depending on certain cases. Using `pytest.mark.parametrize` will be useful I think.

To run the tests, just install tox (pip install tox) and run `tox`. You can give a parameter which the name of the test to skip the others, like `tox murdock/tests/test_job.py` to only the tests that related to the MurdockJob class.

kaspar030 commented:
- changed the "priorities" settings to always have sane defaults (eases accessing them instead of checking for optionals)
- implemented priority label changes (thanks @aabadie for the discussion, much simpler now than my first iteration)
- changed `get_queued_jobs()` to also sort by `job.priority`

aabadie commented:
This needs a rebase!

aabadie commented:
This also needs another rebase, sorry. There are also some failures reported by the CI
